bellman/src/main.rs

```diff
@@ -0,0 +1,315 @@
+#![allow(unused_imports)]
+#![allow(unused_variables)]
+extern crate bellman;
+extern crate pairing;
+extern crate rand;
+extern crate ff;
+use bellman::{Circuit, ConstraintSystem, SynthesisError, Variable, LinearCombination};
+use pairing::{Engine};
+use std::fs::File;
+use std::io::{BufReader, BufRead};
+use std::time::{Instant};
+
+use ff::{
+    Field,
+    PrimeField,
+};
+
+struct ParsedLine<E: Engine> {
+    index: usize,
+    lc: LinearCombination<E>,
+    debug: String,
+    result: E::Fr,
+}
+
+struct Assignment<E: Engine> {
+    variable: Variable,
+    constrained: bool,
+    value: E::Fr,
+}
+
+fn empty_line<E:Engine>(index: usize) -> ParsedLine<E> {
+    ParsedLine::<E>{
+        index:index, 
+        lc: LinearCombination::zero(), 
+        debug: "".to_string(), 
+        result: E::Fr::zero(),
+    }
+}
+
+fn parse_line<E: Engine>(line: &String, variables: &mut Vec<Assignment<E>>, negate: bool) -> ParsedLine<E> {
+    let parts: Vec<&str> = line.split_whitespace().collect();
+    let row: usize = parts[0].parse().unwrap();
+    let column: usize = parts[1].parse().unwrap();
+
+    // Bellman doesn't support parsing negative values.
+    let mut coefficient = E::Fr::from_str(parts[2].trim_left_matches("-")).unwrap();
+    if negate {
+        coefficient.negate();
+    }
+    if parts[2].starts_with("-") { 
+        coefficient.negate();
+    }
+    variables[row] = Assignment::<E>{
+        variable: variables[row].variable, 
+        constrained: true, 
+        value: variables[row].value
+    };
+    let mut result = coefficient;
+    result.mul_assign(&(variables[row].value));
+    ParsedLine::<E> {
+        index: column,
+        lc: LinearCombination::zero() + (coefficient, variables[row].variable),
+        debug: format!("({}*{:?})", coefficient, variables[row].value),
```

Comment:
What is this meant to do? Is this some sort of regex?

Reply:
It's for debugging purposes to print the linear combination in a form that I can reproduce the constraint in case it's not satisfied (e.g. using python)